mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/work/impl/SearchIndexingPlanFilterContextImpl.java

```diff
@@ -0,0 +1,157 @@
+		PojoRawTypeIdentifier<?> typeIdentifier = PojoRawTypeIdentifier.of( clazz );
+		if ( !contextProvider.allNonInterfaceSuperTypes().contains( typeIdentifier ) ) {
+			throw log.unknownClassForNonInterfaceSuperType( clazz, contextProvider.allNonInterfaceSuperTypesClasses() );
+		}
+		return typeIdentifier;
+	}
+
+	public ConfiguredSearchIndexingPlanFilter createFilter(ConfiguredSearchIndexingPlanFilter fallback) {
+		Set<PojoRawTypeIdentifier<?>> allIncludes = new HashSet<>();
+		Set<PojoRawTypeIdentifier<?>> allExcludes = new HashSet<>();
+		Set<PojoRawTypeIdentifier<?>> processedByThisContext = new HashSet<>();
+
+		Set<PojoRawTypeIdentifier<?>> allIndexedAndContainedTypes = contextProvider.allIndexedAndContainedTypes();
+
+		// First we want to go through explicit include/exclude rules defined by this context and
+		// process the hierarchies:
+		for ( PojoRawTypeIdentifier<?> typeIdentifier : contextProvider.allNonInterfaceSuperTypes() ) {
+			if ( includes.contains( typeIdentifier ) ) {
+				include( allIndexedAndContainedTypes, allIncludes, allExcludes, processedByThisContext, typeIdentifier );
+			}
+			else if ( excludes.contains( typeIdentifier ) ) {
+				exclude( allIndexedAndContainedTypes, allIncludes, allExcludes, processedByThisContext, typeIdentifier );
+			}
+		}
+
+		// After context rules are processed we are going through all the types and since `contextProvider.allNonInterfaceSuperTypes()`
+		// includes all possible type identifiers that we care about, we no longer lookup the hierarchies. We only want to
+		// look at the fallback rules for the types that were not processed so far and looking up hierarchies at this point
+		// might lead to an unnecessary collision.
+		// Also at this point we only care about indexed and contained types as we shouldn't get others to be tested
+		// by the filter...
+		for ( PojoRawTypeIdentifier<?> typeIdentifier : allIndexedAndContainedTypes ) {
+			if ( !processedByThisContext.contains( typeIdentifier ) ) {
+				if ( fallback == null || fallback.isIncluded( typeIdentifier ) ) {
+					allIncludes.add( typeIdentifier );
+				}
+				else {
+					allExcludes.add( typeIdentifier );
+				}
+			}
+		}
+
+		return ConfiguredSearchIndexingPlanFilter.create(
+				Collections.unmodifiableSet( allIncludes ),
+				Collections.unmodifiableSet( allExcludes )
+		);
+	}
+
+	private void exclude(Set<PojoRawTypeIdentifier<?>> allIndexedAndContainedTypes,
+			Set<PojoRawTypeIdentifier<?>> allIncludes, Set<PojoRawTypeIdentifier<?>> allExcludes,
+			Set<PojoRawTypeIdentifier<?>> processed, PojoRawTypeIdentifier<?> typeIdentifier) {
+		contextProvider.allByNonInterfaceSuperType( typeIdentifier ).forEach( typeContext -> {
+			PojoRawTypeIdentifier<?> identifier = typeContext.typeIdentifier();
+			// we only care about indexed or contained types. Other types shouldn't be passed to the filter.
+			if ( allIndexedAndContainedTypes.contains( identifier ) ) {
```

Comment:
As far as I can tell, all type contexts refer to either an indexed or a contained type, so this check will always pass and can be removed.